lib/client/weblogin_test.go

```diff
@@ -31,74 +33,257 @@ import (
 	"github.com/gravitational/teleport/lib/client"
 )
 
-func TestPlainHttpFallback(t *testing.T) {
+// TestHostCredentialsHttpFallback tests that HostCredentials requests (/v1/webapi/host/credentials/)
+// fall back to HTTP only if the address is a loopback and the insecure mode was set.
+func TestHostCredentialsHttpFallback(t *testing.T) {
 	testCases := []struct {
-		desc            string
-		path            string
-		handler         http.HandlerFunc
-		actionUnderTest func(ctx context.Context, addr string, insecure bool) error
+		desc     string
+		loopback bool
+		insecure bool
+		fallback bool
 	}{
 		{
-			desc: "HostCredentials",
-			path: "/v1/webapi/host/credentials",
-			handler: func(w http.ResponseWriter, r *http.Request) {
-				if r.RequestURI != "/v1/webapi/host/credentials" {
-					w.WriteHeader(http.StatusNotFound)
-					return
-				}
-				w.Header().Set("Content-Type", "application/json")
-				w.WriteHeader(http.StatusOK)
-				json.NewEncoder(w).Encode(proto.Certs{})
-			},
-			actionUnderTest: func(ctx context.Context, addr string, insecure bool) error {
-				_, err := client.HostCredentials(ctx, addr, insecure, types.RegisterUsingTokenRequest{})
-				return err
-			},
+			desc:     "falls back to http if loopback and insecure",
+			loopback: true,
+			insecure: true,
+			fallback: true,
+		},
+		{
+			desc:     "does not fall back to http if loopback and secure",
+			loopback: true,
+			insecure: false,
+			fallback: false,
+		},
+		{
+			desc:     "does not fall back to http if non-loopback and insecure",
+			loopback: false,
+			insecure: true,
+			fallback: false,
 		},
 	}
 
-	for _, testCase := range testCases {
-		t.Run(testCase.desc, func(t *testing.T) {
-			ctx := context.Background()
+	for _, tc := range testCases {
+		// Start an http server (not https) so that the request only succeeds
+		// if the fallback occurs.
+		var handler http.HandlerFunc = func(w http.ResponseWriter, r *http.Request) {
+			handleRequest(w, r, "/v1/webapi/host/credentials", proto.Certs{})
+		}
+		https := false
+		httpSvr, err := newServer(handler, tc.loopback, https)
+		require.NoError(t, err)
+		defer httpSvr.Close()
+
+		// Send the HostCredentials request.
+		ctx := context.Background()
+		_, err = client.HostCredentials(ctx, httpSvr.Listener.Addr().String(), tc.insecure, types.RegisterUsingTokenRequest{})
+
+		// If it should fallback, then no error should occur
+		// as the request will hit the running http server.
+		if tc.fallback {
+			require.NoError(t, err)
+		} else {
+			require.Error(t, err)
+		}
+	}
+}
 
-			t.Run("Allowed on insecure & loopback", func(t *testing.T) {
-				httpSvr := httptest.NewServer(testCase.handler)
-				defer httpSvr.Close()
+// TestHttpRoundTripperDowngrade tests that the round tripper downgrades https requests to http
+// when HTTP_PROXY is set to "http://localhost:*" (i.e. there's an http proxy running on localhost).
+func TestHttpRoundTripperDowngrade(t *testing.T) {
```

Comment:
I get that this is testing the `HttpRoundTripper` under the hood, however its bad practice to write tests for one package in another. For example, consider if we swap out the underlying transport mechanism in `SSHAgentLogin`. Now this test is bunk, and might just get deleted. At the very least, we would need to rewrite it to test `HTTPRoundTripper` in its own package at that point. 

Another reason this is bad practice is that tests are a form of documentation. If somebody wants to use `HTTPRoundTripper` for something else, it's best that the tests testing its core behavior are close by so they can see how it works.

I would say move these comprehensive tests of `HTTPRoundTripper`'s behavior close to `TestProxyAwareRoundTripper`. And if you think its worthwhile to also have unit tests ensuring this behavior for `client.SSHAgentLogin` and `client.HostCredentials`, then add those to the `client` package.

Reply:
Thanks for the suggestion. Moved the tests in 451646b116a2c646071d235ec78cd9c3717d11d6.